Added lesson01 and lesson01 files

ervbrian opened the pull request:
**Note**
I'm submitting what I have for now, but I am struggling with mocking the main.py functions where input() is used. I attempted to use patching to force return_values, but those lines still show up in the coverage report as untested. Without patching, the functions are called with their respective input() calls, prompting manual input from the user at test runtime. 

**Linting**
```
186590d2efe7:lesson01 ervbrian$ python -m pylint ./inventory_management


Report
======
72 statements analysed.

Statistics by type
------------------

+---------+-------+-----------+-----------+------------+---------+
|type     |number |old number |difference |%documented |%badname |
+=========+=======+===========+===========+============+=========+
|module   |4      |4          |=          |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|class    |3      |3          |=          |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|method   |4      |4          |=          |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+
|function |5      |5          |=          |100.00      |0.00     |
+---------+-------+-----------+-----------+------------+---------+



External dependencies
---------------------
::

    ElectricAppliancesClass (main)
    FurnitureClass (main)
    InventoryClass (ElectricAppliancesClass,FurnitureClass,main)
    market_prices (main)



Raw metrics
-----------

+----------+-------+------+---------+-----------+
|type      |number |%     |previous |difference |
+==========+=======+======+=========+===========+
|code      |97     |53.01 |97       |=          |
+----------+-------+------+---------+-----------+
|docstring |41     |22.40 |41       |=          |
+----------+-------+------+---------+-----------+
|comment   |11     |6.01  |6        |+5.00      |
+----------+-------+------+---------+-----------+
|empty     |34     |18.58 |32       |+2.00      |
+----------+-------+------+---------+-----------+



Duplication
-----------

+-------------------------+------+---------+-----------+
|                         |now   |previous |difference |
+=========================+======+=========+===========+
|nb duplicated lines      |0     |0        |=          |
+-------------------------+------+---------+-----------+
|percent duplicated lines |0.000 |0.000    |=          |
+-------------------------+------+---------+-----------+



Messages by category
--------------------

+-----------+-------+---------+-----------+
|type       |number |previous |difference |
+===========+=======+=========+===========+
|convention |0      |0        |=          |
+-----------+-------+---------+-----------+
|refactor   |0      |0        |=          |
+-----------+-------+---------+-----------+
|warning    |0      |0        |=          |
+-----------+-------+---------+-----------+
|error      |0      |0        |=          |
+-----------+-------+---------+-----------+




--------------------------------------------------------------------
Your code has been rated at 10.00/10 (previous run: 10.00/10, +0.00)
```

**Unit Testing**
```
186590d2efe7:lesson01 ervbrian$ PYTHONPATH=inventory_management/ python -m coverage run --source=inventory_management -m unittest tests/test_unit.py && python -m coverage report -m
----------------------------------------------------------------------
Ran 10 tests in 0.003s

OK
Name                                              Stmts   Miss  Cover   Missing
-------------------------------------------------------------------------------
inventory_management/ElectricAppliancesClass.py       6      0   100%
inventory_management/FurnitureClass.py                6      0   100%
inventory_management/InventoryClass.py                8      0   100%
inventory_management/main.py                         52     20    62%   67-69, 75-89, 97-103, 116-119
inventory_management/market_prices.py                 2      0   100%
-------------------------------------------------------------------------------
TOTAL                                                74     20    73%
```

nairraghav commented:
Hey Brian, when I pulled down your code, I'm seeing different numbers in coverage. Do you mind double-checking that you'll pushed all your code?

ervbrian commented:
> Hey Brian, when I pulled down your code, I'm seeing different numbers in coverage. Do you mind double-checking that you'll pushed all your code?

Updated the overview with correct coverage details.

nairraghav commented:
Looks like the pylint issue was on my end, I resolved it and was able to see it rate a 10!

nairraghav commented on students/BrianE/lesson01/inventory_management/ElectricAppliancesClass.py:
```diff
@@ -0,0 +1,18 @@
+""" ElectricAppliances Class """
+
+
```
Since we are running our tests from the inventory_management folder, you'll want to modify your imports to be pulled from there. Ex:
from inventory_management.InventoryClass import Inventory

nairraghav commented on students/BrianE/lesson01/inventory_management/FurnitureClass.py:
```diff
@@ -0,0 +1,18 @@
+""" Furniture Class """
+
+
+from InventoryClass import Inventory
```
Same import fix here

nairraghav commented on students/BrianE/lesson01/inventory_management/main.py:
```diff
@@ -0,0 +1,119 @@
+""" Launches the user interface for the inventory management system """
+
+import sys
+import market_prices
```
Same import fixes here as well

nairraghav commented on students/BrianE/lesson01/tests/test_unit.py:
```diff
@@ -0,0 +1,146 @@
+            exit_program(full_inventory={})
+
+        with self.assertRaises(TypeError):
+            exit_program()
```
Great job with the tests!

You should be able to do multiple mock patches like the following: https://gist.githubusercontent.com/yeraydiazdiaz/25d95ff91789d46cb45aa2cb6a7c2dac/raw/dfa0072f4558cea2f978e3c97eee90fdb7e2a2db/wtmock_04_builtins_decorators.py


As discussed, I will put out a video walking through an example

nairraghav commented on students/BrianE/lesson02/assignment/charges_calc.py:
```diff
@@ -0,0 +1,152 @@
+        try:
+            data = json.load(file)
+            logger.debug(f'Successfully loaded {len(data.keys())} elements from {filename}')
+        except Exception as e:
```
While using a generic Exception here will solve all your problems, we should try to use specific Exception catches so that the User is aware of the incorrect behavior that is causing the problem. For example, this could be because the file doesn't exist or due to the json decoder failing.

nairraghav commented on students/BrianE/lesson02/assignment/charges_calc.py:
```diff
@@ -0,0 +1,152 @@
+
+        # if start date is before end date, do not process
+        if value['rental_start'] > value['rental_end']:
+            logger.error('start date before end date. Skipping...')
```
This is one way to bypass! We could also try to help the system by using absolute value to always get a positive number of days.

nairraghav commented:
Well done Brian, this looks great! Please take a look at my comments for some small fixes / tips but overall you did an awesome job.

I am not going to merge this in yet because it looks like you removed one of the template files: students/template_student/lesson02/assignment/src/charges_calc.py

To fix this, you'll need to do multiple things:

Set the remote upstream to the original repo:
git remote add upstream https://github.com/UWPCE-PythonCert-ClassRepos/220-Advanced-Summer-2019.git

Replace the file: 
git checkout upstream/master -- students/template_student/lesson02/assignment/src/charges_calc.py

Once you have fixed this, I will merge this!


Let me know if you have any questions.

ervbrian commented:
> Well done Brian, this looks great! Please take a look at my comments for some small fixes / tips but overall you did an awesome job.
> 
> I am not going to merge this in yet because it looks like you modified some of the template files: deleted - students/template_student/lesson02/assignment/src/charges_calc.py
> renamed/edited - students/template_student/lesson02/assignment/src/source.json
> 
> To fix this, you'll need to do multiple things:
> 
> Set the remote upstream to the original repo:
> git remote add upstream https://github.com/UWPCE-PythonCert-ClassRepos/220-Advanced-Summer-2019.git
> 
> Replace the file:
> git checkout upstream/master -- students/template_student/lesson02/assignment/src/charges_calc.py
> git checkout upstream/master -- students/template_student/lesson02/assignment/src/source.json
> 
> Once you have fixed this, I will merge this! Let me know if you have any questions.

Whoops. Added back in.

nairraghav commented:
Hey Brian,

I saw that you added one of the files back but could you double check that the source.json file exists in the template student?

students/template_student/lesson02/assignment/src/source.json


It looks like this file still shows up as a rename so I'm wondering if you only have your copy (overwriting the existing copy)

ervbrian commented:
> Hey Brian,
> 
> I saw that you added one of the files back but could you double check that the source.json file exists in the template student?
> 
> students/template_student/lesson02/assignment/src/source.json
> 
> It looks like this file still shows up as a rename so I'm wondering if you only have your copy (overwriting the existing copy)

Added this back in as well.

nairraghav commented on students/BrianE/lesson03/customer_data/basic_operations.py:
```diff
@@ -0,0 +1,125 @@
+            customer.save()
+        return True
+    except Exception:  # return empty dictionary if customer_id doesn't match
+        return {}
```
Since you return True in the successful deletion, consider returning False here so that your returns make sense (having one as True and one as an Empty Dict isn't as intuitive)

nairraghav commented on students/BrianE/lesson03/customer_data/basic_operations.py:
```diff
@@ -0,0 +1,125 @@
+            customer = Customer.get(Customer.customer_id == customer_id)
+            customer.update(Customer.credit_limit == credit_limit)
+            customer.save()
+    except Exception as e:
```
Similar to your search for customer, you may want to have a specific handling of a customer_id that doesn't exist

nairraghav commented on students/BrianE/lesson03/tests/test_gradel03.py:
```diff
@@ -0,0 +1,173 @@
+                       )
+    actives = l.list_active_customers()
+
+    assert actives == 4
```
I believe this would need to be updated to assert len(actives) == 4 since you are comparing how many active accounts there are

nairraghav commented on students/BrianE/lesson03/tests/test_gradel03.py:
```diff
@@ -0,0 +1,173 @@
+
+    for customer in _list_active_customers:
+        l.delete_customer(customer[0])
+
```
It would be a good idea to add another assert here to validate they were all deleted:
assert len(actives) == 0

nairraghav commented on students/BrianE/lesson03/tests/test_gradel03.py:
```diff
@@ -0,0 +1,173 @@
+                       )
+
+        response = l.delete_customer(customer[0])
+        assert response is True
```
Since response is True, you can just do "assert response"

nairraghav commented on students/BrianE/lesson03/tests/test_gradel03.py:
```diff
@@ -0,0 +1,173 @@
+        assert response is True
+
+        deleted = l.search_customer(customer[0])
+        assert deleted == {}
```
If you change the code to return false instead, you can just do "assert not deleted"

nairraghav commented on students/BrianE/lesson04/basic_operations.py:
```diff
@@ -0,0 +1,181 @@
+            customer.save()
+        logger.info(f"Customer {customer_id} deleted")
+        return True
+    except Exception as e:  # return empty dictionary if customer_id doesn't match
```
You  can use the same Exception logic as above to first try to catch CustomerDoesNotExist

nairraghav commented on students/BrianE/lesson04/basic_operations.py:
```diff
@@ -0,0 +1,181 @@
+            customer.update(Customer.credit_limit == credit_limit)
+            customer.save()
+        logger.info(f"Credit limit updated for customer {customer_id}")
+    except Exception as e:
```
Similarly, use the right exception to catch instead of the catch all

nairraghav commented on students/BrianE/lesson04/tests/test_gradel03.py:
```diff
@@ -0,0 +1,173 @@
+                       )
+    actives = l.list_active_customers()
+
+    assert actives == 4
```
As before, you will want to assert len(actives) == 4

nairraghav commented:
Great work Brian, please take a look at my comments for some small changes that you could implement for better behavior but well done overall!
